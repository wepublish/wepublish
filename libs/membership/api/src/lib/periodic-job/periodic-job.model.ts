import {Field, ObjectType} from '@nestjs/graphql'

@ObjectType()
export class PeriodicJob {
  @Field()
  id!: string

  @Field()
  createdAt!: Date

  @Field()
  modifiedAt!: Date

  @Field()
  date!: Date

  @Field()
  tries!: number

  @Field({nullable: true})
  executionTime?: Date

  @Field({nullable: true})
  successfullyFinished?: Date

  @Field({nullable: true})
  finishedWithError?: Date

  @Field({nullable: true})
  error?: string
}
