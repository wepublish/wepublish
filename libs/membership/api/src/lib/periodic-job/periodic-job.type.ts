export type PeriodicJobRunObject = {
  date: Date
  isRetry: boolean
}
