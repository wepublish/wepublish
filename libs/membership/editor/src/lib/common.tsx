import {ApolloClient, ApolloError, NormalizedCacheObject} from '@apollo/client'
import {TFunction} from 'i18next'
import {Message, toaster} from 'rsuite'

const showErrors = (error: ApolloError): void => {
  toaster.push(
    <Message type="error" showIcon closable duration={3000}>
      {error.message}
    </Message>
  )
}

const showSavedToast = (t: TFunction): void => {
  toaster.push(
    <Message type="success" showIcon closable duration={3000}>
      {t('subscriptionFlow.savedChange').toString()}
    </Message>
  )
}

/**
 * Default options for the GraphQL client. Displays errors and a completion message.
 * @param client the graphql client to make the request with
 * @param t the translation instance
 * @returns QueryHookOptions for the GraphQL client
 */
export const DEFAULT_MUTATION_OPTIONS = (
  client: ApolloClient<NormalizedCacheObject>,
  t: TFunction
) => {
  return {
    ...DEFAULT_QUERY_OPTIONS(client),
    onCompleted: () => showSavedToast(t)
  }
}

/**
 * Default options for the GraphQL client. Displays errors.
 * @param client the graphql client to make the request with
 * @returns QueryHookOptions for the GraphQL client
 */
export const DEFAULT_QUERY_OPTIONS = (client: ApolloClient<NormalizedCacheObject>) => {
  return {
    client,
    onError: showErrors
  }
}
