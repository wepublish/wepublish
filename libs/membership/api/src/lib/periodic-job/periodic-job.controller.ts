import {OldContextService, PrismaService} from '@wepublish/nest-modules'
import {add, addDays, differenceInDays, endOfDay, set, startOfDay, sub, subMinutes} from 'date-fns'
import {SubscriptionEventDictionary} from '../subscription-event-dictionary/subscription-event-dictionary'
import {
  Invoice,
  InvoiceItem,
  MemberPlan,
  PaymentMethod,
  PaymentProviderCustomer,
  PeriodicJob,
  Subscription,
  SubscriptionEvent,
  SubscriptionPeriod,
  User
} from '@prisma/client'
import {PeriodicJobRunObject} from './periodic-job.type'
import {Injectable, Logger} from '@nestjs/common'
import {MailController, mailLogType} from '../mail/mail.controller'
import {Action} from '../subscription-event-dictionary/subscription-event-dictionary.type'
import {SubscriptionController} from '../subscription/subscription.controller'
const FIVE_MINUTES_IN_MS = 5 * 60 * 1000

/**
 * Controller responsible for performing periodic jobs. A new controller
 * instance must be created for every run.
 */
@Injectable()
export class PeriodicJobController {
  private subscriptionEventDictionary: SubscriptionEventDictionary
  private runningJob: PeriodicJob | undefined
  private readonly logger = new Logger('PeriodicJobController')
  private randomNumberRangeForConcurrency = FIVE_MINUTES_IN_MS
  constructor(
    private readonly prismaService: PrismaService,
    private readonly oldContextService: OldContextService,
    private readonly subscriptionController: SubscriptionController
  ) {
    this.subscriptionEventDictionary = new SubscriptionEventDictionary(this.prismaService)
  }

  /**
   * Run the periodic jobs. This makes sure that no two instances of the same
   * controller run their jobs at the same time and returns if they are.
   * @returns void
   */
  public async concurrentExecute() {
    await this.sleepForRandomIntervalToEnsureConcurrency()
    if (await this.isAlreadyAJobRunning()) {
      this.logger.log('Periodic job already running on an other instance. skipping...')
      return
    }
    await this.execute()
  }

  /**
   * Runs all outstanding {@link getOutstandingRuns} runs by doing the following for each run:
   * - send custom mails
   * - create invoices
   * - charge due invoices
   * - deactivate overdue subscriptions
   * If any of the tasks fail, the entire job is marked as failed.
   */
  public async execute() {
    await this.loadEnvironment()
    for (const periodicJobRunObject of await this.getOutstandingRuns()) {
      if (periodicJobRunObject.isRetry) {
        await this.retryFailedJob(periodicJobRunObject.date)
      } else {
        await this.markJobStarted(periodicJobRunObject.date)
      }
      try {
        await this.findAndSendCustomMails(periodicJobRunObject)
        await this.findAndCreateInvoices(periodicJobRunObject)
        await this.findAndChargeDueInvoices(periodicJobRunObject)
        await this.findAndDeactivateSubscriptions(periodicJobRunObject)
      } catch (e) {
        await this.markJobFailed((e as Error).toString())
        throw e
      }
      await this.markJobSuccessful()
    }
  }

  private async loadEnvironment() {
    await this.subscriptionEventDictionary.initialize()
  }

  private async findAndDeactivateSubscriptions(
    periodicJobRunObject: PeriodicJobRunObject
  ): Promise<void> {
    const unpaidInvoices = await this.subscriptionController.getSubscriptionsToDeactivate(
      periodicJobRunObject.date
    )
    for (const unpaidInvoice of unpaidInvoices) {
      await this.deactivateSubscription(periodicJobRunObject, unpaidInvoice)
    }
  }

  private async findAndChargeDueInvoices(periodicJobRunObject: PeriodicJobRunObject) {
    const invoicesToCharge = await this.subscriptionController.getInvoicesToCharge(
      endOfDay(periodicJobRunObject.date)
    )
    for (const invoiceToCharge of invoicesToCharge) {
      await this.chargeInvoice(periodicJobRunObject, invoiceToCharge)
    }
  }

  private async findAndCreateInvoices(periodicJobRunObject: PeriodicJobRunObject) {
    const subscriptionsToCreateInvoice =
      await this.subscriptionController.getSubscriptionsForInvoiceCreation(
        periodicJobRunObject.date,
        this.subscriptionEventDictionary.getEarliestInvoiceCreationDate(periodicJobRunObject.date)
      )
    for (const subscriptionToCreateInvoice of subscriptionsToCreateInvoice) {
      await this.createInvoice(periodicJobRunObject, subscriptionToCreateInvoice)
    }
  }

  private async findAndSendCustomMails(periodicJobRunObject: PeriodicJobRunObject) {
    const subscriptionsWithEvents = await this.prismaService.subscription.findMany({
      where: {
        OR: this.subscriptionEventDictionary
          .getDatesWithCustomEvent(periodicJobRunObject.date)
          .map(date => ({
            paidUntil: {
              gte: date,
              lte: subMinutes(
                set(date, {hours: 23, minutes: 59, seconds: 59, milliseconds: 999}),
                date.getTimezoneOffset()
              )
            }
          }))
      },
      include: {
        user: true
      }
    })
    for (const subscriptionsWithEvent of subscriptionsWithEvents) {
      await this.sendCustomMails(periodicJobRunObject, subscriptionsWithEvent)
    }
  }

  private async sendCustomMails(
    periodicJobRunObject: PeriodicJobRunObject,
    subscriptionsWithEvent: Subscription & {user: User}
  ) {
    const daysAwayFromEnding = differenceInDays(
      periodicJobRunObject.date,
      startOfDay(subscriptionsWithEvent.paidUntil!)
    )
    const subscriptionDictionary = this.subscriptionEventDictionary.getActionFromStore({
      memberplanId: subscriptionsWithEvent.memberPlanID,
      paymentmethodeId: subscriptionsWithEvent.paymentMethodID,
      periodicity: subscriptionsWithEvent.paymentPeriodicity,
      autorenwal: subscriptionsWithEvent.autoRenew,
      daysAwayFromEnding
    })

    for (const event of subscriptionDictionary) {
      if (event.type === SubscriptionEvent.CUSTOM) {
        await this.sendTemplateMail(
          event,
          subscriptionsWithEvent.user,
          periodicJobRunObject.isRetry,
          subscriptionsWithEvent,
          periodicJobRunObject.date
        )
      }
    }
  }

  private async createInvoice(
    periodicJobRunObject: PeriodicJobRunObject,
    subscriptionToCreateInvoice: Subscription & {
      periods: SubscriptionPeriod[]
      user: User
      memberPlan: MemberPlan
    }
  ) {
    const eventInvoiceCreation = this.subscriptionEventDictionary.getActionFromStore({
      memberplanId: subscriptionToCreateInvoice.memberPlanID,
      paymentmethodeId: subscriptionToCreateInvoice.paymentMethodID,
      periodicity: subscriptionToCreateInvoice.paymentPeriodicity,
      autorenwal: subscriptionToCreateInvoice.autoRenew,
      events: [SubscriptionEvent.INVOICE_CREATION, SubscriptionEvent.DEACTIVATION_UNPAID]
    })
    const creationEvent = eventInvoiceCreation.find(
      e => e.type === SubscriptionEvent.INVOICE_CREATION
    )
    if (!creationEvent) {
      throw new Error('No invoice creation found!')
    }
    const deactivationEvent = eventInvoiceCreation.find(
      e => e.type === SubscriptionEvent.DEACTIVATION_UNPAID
    )
    if (!deactivationEvent) {
      throw new Error('No invoice deactivation event found!')
    }

    if (
      subscriptionToCreateInvoice.paidUntil &&
      add(startOfDay(subscriptionToCreateInvoice.paidUntil), {
        days: creationEvent.daysAwayFromEnding!
      }) > periodicJobRunObject.date
    ) {
      return false
    }

    await this.subscriptionController.createInvoice(subscriptionToCreateInvoice, deactivationEvent)

    await this.sendTemplateMail(
      creationEvent,
      subscriptionToCreateInvoice.user,
      periodicJobRunObject.isRetry,
      subscriptionToCreateInvoice,
      periodicJobRunObject.date
    )
    return true
  }

  private async chargeInvoice(
    periodicJobRunObject: PeriodicJobRunObject,
    invoiceToCharge: Invoice & {
      subscription:
        | (Subscription & {
            paymentMethod: PaymentMethod
            memberPlan: MemberPlan
            user: (User & {paymentProviderCustomers: PaymentProviderCustomer[]}) | null
          })
        | null
      items: InvoiceItem[]
      subscriptionPeriods: SubscriptionPeriod[]
    }
  ) {
    if (!invoiceToCharge.subscription) {
      throw new Error(`Invoice ${invoiceToCharge.id} has no subscription assigned!`)
    }

    const eventsRenewal = this.subscriptionEventDictionary.getActionFromStore({
      memberplanId: invoiceToCharge.subscription.memberPlanID,
      paymentmethodeId: invoiceToCharge.subscription.paymentMethodID,
      periodicity: invoiceToCharge.subscription.paymentPeriodicity,
      autorenwal: invoiceToCharge.subscription.autoRenew,
      events: [SubscriptionEvent.RENEWAL_SUCCESS, SubscriptionEvent.RENEWAL_FAILED]
    })

    const mailAction = await this.subscriptionController.chargeInvoice(
      invoiceToCharge,
      eventsRenewal
    )

    if (mailAction.action) {
      const user = Object.assign({}, invoiceToCharge.subscription.user)
      await this.sendTemplateMail(
        mailAction.action,
        user,
        periodicJobRunObject.isRetry,
        {
          errorCode: mailAction.errorCode,
          ...invoiceToCharge
        },
        periodicJobRunObject.date
      )
    }
  }

  private async deactivateSubscription(
    periodicJobRunObject: PeriodicJobRunObject,
    unpaidInvoice: Invoice & {subscription: (Subscription & {user: User}) | null}
  ) {
    if (!unpaidInvoice.subscription) {
      throw new Error(`Invoice ${unpaidInvoice.id} has no subscription assigned!`)
    }
    const eventDeactivationUnpaid = this.subscriptionEventDictionary.getActionFromStore({
      memberplanId: unpaidInvoice.subscription.memberPlanID,
      paymentmethodeId: unpaidInvoice.subscription.paymentMethodID,
      periodicity: unpaidInvoice.subscription.paymentPeriodicity,
      autorenwal: unpaidInvoice.subscription.autoRenew,
      events: [SubscriptionEvent.DEACTIVATION_UNPAID]
    })
    if (!eventDeactivationUnpaid[0]) {
      throw new Error('No subscription deactivation found!')
    }

    await this.subscriptionController.deactivateSubscription(unpaidInvoice)

    await this.sendTemplateMail(
      eventDeactivationUnpaid[0],
      unpaidInvoice.subscription.user,
      periodicJobRunObject.isRetry,
      unpaidInvoice,
      periodicJobRunObject.date
    )
  }

  /**
   * Mark a job as re-trying at the current date.
   * @param runDate The original date of the job run.
   */
  private async retryFailedJob(runDate: Date) {
    this.runningJob = await this.prismaService.periodicJob.update({
      where: {
        date: runDate
      },
      data: {
        executionTime: new Date()
      }
    })
    this.logger.warn('Retry failed job!')
  }

  /**
   * Mark a job as started at the current date.
   * @param runDate the original date of the job run.
   */
  private async markJobStarted(runDate: Date) {
    this.runningJob = await this.prismaService.periodicJob.create({
      data: {
        date: runDate,
        executionTime: new Date()
      }
    })
  }

  /**
   * Check if any job is already being processed.
   * @returns if there are any jobs running.
   */
  private async isAlreadyAJobRunning(): Promise<boolean> {
    const runLimit = sub(new Date(), {hours: 2})
    const runs = await this.prismaService.periodicJob.findMany({
      where: {
        executionTime: {
          gte: runLimit
        }
      }
    })
    return runs.length > 0
  }

  /**
   * Mark a job as completed in the database.
   */
  private async markJobSuccessful() {
    if (!this.runningJob) {
      throw new Error('Try to make a job as successful while none is running!')
    }
    await this.prismaService.periodicJob.update({
      where: {
        id: this.runningJob.id
      },
      data: {
        successfullyFinished: new Date(),
        tries: ++this.runningJob.tries
      }
    })
    this.runningJob = undefined
  }

  /**
   * Sleep for a random time between 0 and 300 seconds to ensure that two parallel processes
   * are not starting to process the queue at the same time.
   * @returns void
   */
  private async sleepForRandomIntervalToEnsureConcurrency() {
    const randomSleepTimeout = Math.floor(Math.random() * this.randomNumberRangeForConcurrency)
    this.logger.log(
      `To ensure concurrent execution in multi instance environment choosing random number between 0 and ${this.randomNumberRangeForConcurrency}... sleeping for  ${randomSleepTimeout}ms`
    )
    const sleep = (ms: number) => new Promise(r => setTimeout(r, ms))
    await sleep(randomSleepTimeout)
    return randomSleepTimeout
  }

  /**
   * Mark a job as failed in the database by incrementing the `tries` count and updating the failure timestamp.
   * @param error a description of the error
   */
  private async markJobFailed(error: string) {
    if (!this.runningJob) {
      throw new Error('Try to make a job as failed while none is running!')
    }
    await this.prismaService.periodicJob.update({
      where: {
        id: this.runningJob.id
      },
      data: {
        finishedWithError: new Date(),
        tries: ++this.runningJob.tries,
        error
      }
    })
    this.runningJob = undefined
  }

  /**
   * Calculate the runs in the past that have not completed yet.
   * - If the Controller is run for the first time, this returns just todays run.
   * - If the last run had an error, it returns the run when the error happened.
   * - If there was an execution pause, it returns all runs between the last successful run and the current day.
   * @returns An array of pending runs.
   */
  private async getOutstandingRuns(): Promise<PeriodicJobRunObject[]> {
    const today = new Date()
    const runDates: PeriodicJobRunObject[] = []
    const latestRun = await this.prismaService.periodicJob.findFirst({
      orderBy: {
        date: 'desc'
      }
    })
    if (!latestRun) {
      this.logger.debug('Periodic job first run')
      return [{isRetry: false, date: startOfDay(today)}]
    }
    if (latestRun.finishedWithError && !latestRun.successfullyFinished) {
      this.logger.warn('Last run had errors retrying....')
      runDates.push({isRetry: true, date: startOfDay(latestRun.date)})
    }
    return runDates.concat(this.generateDateArray(latestRun.date, today))
  }

  /**
   * Generate an array of dates between the two bounds
   * @param startDate The beginning date (inclusive)
   * @param endDate The ending date (exclusive)
   * @returns An array of Date objects
   */
  private generateDateArray(startDate: Date, endDate: Date) {
    const dateArray = []
    const lastDate = startOfDay(endDate)
    let inputDate = startOfDay(startDate)
    while (inputDate < lastDate) {
      inputDate = addDays(inputDate, 1)
      dateArray.push({isRetry: false, date: inputDate})
    }
    return dateArray
  }

  /**
   * Send an email and store it in the Mail Log
   * @param action the event and template to send
   * @param user the recipient
   * @param isRetry whether this is a retried delivery
   * @param optionalData unknown
   * @param periodicJobRunDate the current date for the delivery
   */
  private async sendTemplateMail(
    action: Action,
    user: User,
    isRetry: boolean,
    optionalData: Record<string, any>,
    periodicJobRunDate: Date
  ) {
    if (action.externalMailTemplate && user) {
      await new MailController(this.prismaService, this.oldContextService, {
        daysAwayFromEnding: action.daysAwayFromEnding,
        externalMailTemplateId: action.externalMailTemplate,
        recipient: user,
        isRetry,
        optionalData,
        periodicJobRunDate,
        mailType: mailLogType.SubscriptionFlow
      }).sendMail()
    }
  }
}
