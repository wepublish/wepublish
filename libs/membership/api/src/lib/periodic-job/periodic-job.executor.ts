import {Cron} from '@nestjs/schedule'
import {Injectable} from '@nestjs/common'
import {PeriodicJobController} from './periodic-job.controller'

const SCHEDULE = process.env.PERIODIC_JOB_EXECUTION_SCHEDULE || '0 0 3 * * *'

@Injectable()
export class PeriodicJobExecutor {
  constructor(private readonly periodicJobController: PeriodicJobController) {}

  @Cron(
    SCHEDULE,
    // Allow only chron that run once a day example: [number] [number] [number] [star] [star] [star]
    {
      disabled: !/[1-5]?[0-9] [1-5]?[0-9] [1-2]?[0-9] \* \* \*/.test(SCHEDULE)
    }
  )
  async handleCron() {
    await this.periodicJobController.concurrentExecute()
  }
}
