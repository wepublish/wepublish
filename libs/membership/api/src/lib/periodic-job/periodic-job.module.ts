import {Module} from '@nestjs/common'
import {PeriodicJobExecutor} from './periodic-job.executor'
import {OldContextModule, PrismaModule} from '@wepublish/nest-modules'
@Module({
  controllers: [],
  providers: [PeriodicJobExecutor],
  exports: [],
  imports: [PrismaModule, OldContextModule]
})
export class PeriodicJobModule {}
