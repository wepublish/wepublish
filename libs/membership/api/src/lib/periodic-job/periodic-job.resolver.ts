import {Args, Int, Query, Resolver} from '@nestjs/graphql'
import {CanGetPeriodicJobLog, Permissions} from '@wepublish/permissions/api'
import {PeriodicJob} from './periodic-job.model'
import {PeriodicJobService} from './periodic-job.service'

@Resolver(() => PeriodicJob)
export class PeriodicJobResolver {
  constructor(private periodicJobService: PeriodicJobService) {}

  @Permissions(CanGetPeriodicJobLog)
  @Query(() => [PeriodicJob])
  periodicJobLog(
    @Args('take', {type: () => Int, nullable: true, defaultValue: 10}) take: number,
    @Args('skip', {type: () => Int, nullable: true}) skip?: number
  ) {
    return this.periodicJobService.getJobLog(take, skip)
  }
}
