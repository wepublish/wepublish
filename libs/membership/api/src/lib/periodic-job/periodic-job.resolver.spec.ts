import {INestApplication, Module} from '@nestjs/common'
import request from 'supertest'
import {GraphQLModule} from '@nestjs/graphql'
import {ApolloDriverConfig, ApolloDriver} from '@nestjs/apollo'
import {PrismaModule} from '@wepublish/nest-modules'
import {Test, TestingModule} from '@nestjs/testing'
import {PeriodicJobResolver} from './periodic-job.resolver'
import {PeriodicJobService} from './periodic-job.service'
import {PeriodicJob} from './periodic-job.model'

@Module({
  imports: [
    GraphQLModule.forRoot<ApolloDriverConfig>({
      driver: ApolloDriver,
      autoSchemaFile: true,
      path: '/',
      cache: 'bounded'
    }),
    PrismaModule
  ],
  providers: [PeriodicJobResolver, {provide: PeriodicJobService, useValue: {getJobLog: jest.fn()}}]
})
export class AppModule {}

const periodicJobsQuery = `
    query PeriodicJobLogs($skip: Int, $take: Int) {
        periodicJobLog(skip: $skip, take: $take) {
            id
            tries
            executionTime
            finishedWithError
            successfullyFinished
        }
    }
`

export const mockLogs: PeriodicJob[] = [
  {
    id: '1234',
    createdAt: new Date('2023-01-01'),
    modifiedAt: new Date('2023-01-01'),
    date: new Date('2023-01-01'),
    tries: 1,
    executionTime: new Date('2023-01-01'),
    finishedWithError: new Date('2023-01-01'),
    successfullyFinished: new Date('2023-01-01')
  }
]

describe('ConsentResolver', () => {
  let app: INestApplication
  let service: PeriodicJobService

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [AppModule]
    }).compile()

    service = module.get<PeriodicJobService>(PeriodicJobService)
    app = module.createNestApplication()
    await app.init()
  })

  afterAll(async () => {
    await app.close()
  })

  it('periodic jobs query', async () => {
    const spy = jest.spyOn(service, 'getJobLog').mockReturnValue(Promise.resolve(mockLogs) as any)

    await request(app.getHttpServer())
      .post('')
      .send({
        query: periodicJobsQuery,
        variables: {
          take: 1
        }
      })
      .expect(res => {
        expect(res.body.data.periodicJobLog).toMatchSnapshot()
        expect(spy).toHaveBeenCalledWith(1, undefined)
      })
      .expect(200)
  })
})
