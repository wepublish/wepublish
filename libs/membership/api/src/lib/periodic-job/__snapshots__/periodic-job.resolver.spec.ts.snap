// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsentResolver periodic jobs query 1`] = `
[
  {
    "executionTime": "2023-01-01T00:00:00.000Z",
    "finishedWithError": "2023-01-01T00:00:00.000Z",
    "id": "1234",
    "successfullyFinished": "2023-01-01T00:00:00.000Z",
    "tries": 1,
  },
]
`;
